README.md
# ELS - Enterprise Linux/EuroLinux Sources - tcl-pgtcl
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/NEWS
This is the pgtcl-ng Source Release NEWS file
Last updated for pgtclng-2.0.0 on 2011-09-18
The project home page is: http://sourceforge.net/projects/pgtclng/
-----------------------------------------------------------------------------

* 2011-09-18    pgtcl-ng 2.0.0 Released

This release includes two new commands and one new command option. It also
includes a fix for building with Tcl-8.6 (currently in beta).

This is the first release after rehosting on Sourceforge.net. Documentation
and information files have been updated accordingly.

Feature request #3408919 "Access to server PID, notification PID, and
        server version":
+ New command pg_backend_pid to get the backend process ID.
+ New command pg_server_version to get the server version as an intger.
+ New -pid option to pg_listen, to pass the notifying client's backend
  process ID to the notification callback.

Fixed bug #3405928, building fails with Tcl-8.6 beta, reported by Zbigniew:
Pgtcl-ng will now compile with Tcl-8.6beta, avoiding access to the errorLine
structure element which is now hidden.  The correct access method is detected
at compile time. However, if Pgtcl-ng is compiled using stubs with Tcl-8.6,
and then run under Tcl-8.5 or older, it will crash. This is because the
new function used at 8.6 to access the structure member does not exist in
older releases.


Note: Despite the version number ending in ".0.0", Pgtclng-2.0.0 is not 
substantially different from the previous release 1.9.0, and there should
be no incompatibilities. (Changes were made to an internal data structure,
but a major factor in not using "1.10.0" instead is the way release packages
are sorted by name.)


* 2011-03-21    pgtcl-ng 1.9.0 Released

This release adds one new feature: pg_result $r -dict, which returns the
query result as a Tcl dictionary. The idea for this feature came from the
pgfoundry 'pgtcl' project. This feature requires Tcl-8.5 or higher.


* 2010-10-10    pgtcl-ng 1.8.0 Released

This release has one new feature for PostgreSQL-9.0.0. Notification
messages can now include a payload, which is passed to the notification
listener callback proc. For example:
Given (in one session):
    pg_listen $db my_channel my_callback_proc

Then (possibly in another session):
    SQL> NOTIFY my_channel, 'the payload'
This will result in execution of:  my_callback_proc "the payload" in the
original session.

And:
       SQL> NOTIFY my_channel
 or:   SQL> NOTIFY my_channel, ''
This will result in execution of:  my_callback_proc
in the original session.

Compatibility Warning:
    This applies only if you use pg_listen to set up a notification listener
    callback procedure.

    Your listener callback should be defined to accept an optional argument
    for the payload, for example:   proc listen_handler {{payload ""}} { ... }
    Starting with version 1.8.0, pgtclng will pass a payload argument to the
    handler if a non-empty payload is provided in the SQL command. If an empty
    payload is provided, or no payload (including any usage with a PostgreSQL
    server older than 9.0.0), pgtclng will not supply the argument to the
    handler. This behavior was chosen to improve compatibility with older
    scripts that would throw an error if provided an unexpected argument.

    If you do not update your listener callback to have an optional argument,
    and you never include a payload in the notification SQL, your script will
    not have any problems. However, note that anyone who can connect to the
    database can send a notification (if they know the 'channel' name used
    in the pg_listen command), and they can include a payload. If your listener
    callback does not expect a payload argument, it will throw a background
    error (which may or may not terminate the script) if it receives such a
    payload argument.


For Windows, only MinGW building is now supported and tested. Borland BCC
building of pgtcl-ng probably still works, but is no longer tested. Refer
to the Windows binary release of pgtcl-ng for more information.


* 2010-07-14    pgtcl-ng 1.7.2 Released

This release now supports building on Windows with MinGW, in addition to
Borland C++ Building (BCC). The binary release for Windows now includes
DLLs built with BCC and with MinGW. (In future releases, the BCC-built DLLs
may be dropped, leaving just MinGW.)

There is one bug fix for a failure to build and run pgtcl-ng on PostgreSQL
versions before 8.2. It has no impact for PostgreSQL-8.2 and higher.


* 2009-11-25    pgtcl-ng 1.7.1 Released

The release now supports building a Tcl stubs-enabled module on Windows,
which means the same compiled module will load into different Tcl versions.

There is one minor bug fix to pg_encrypt_password. An error was not
properly detected when the command was passed an invalid, empty username.


* 2009-09-11    pgtcl-ng 1.7.0 Released

This release adds some commands and options to "catch up" to features added
to the PostgreSQL libpq library.

This version adds four new commands:
+ pg_encrypt_password to encrypt a password for certain SQL commands
+ pg_lo_truncate to truncate a large object
+ pg_describe_cursor to return information about a cursor (portal)
+ pg_describe_prepared to return information about a prepared statement.

This version adds two options to pg_result, for use with
pg_describe_prepared to return information about a prepared statement:
+ pg_result -numParams returns the number of parameters
+ pg_result -paramTypes returns the parameter type OIDs.

In this release, pg_escape_string, pg_quote, and pg_escape_bytea
accept an optional connection parameter. This allows use of
connection-specific information to properly handle string escaping.

This release also contains fixes to error handling in pg_select.


* 2007-12-26    pgtcl-ng 1.6.2 Released

This release fixes a character set encoding problem on Windows only. When
communicating with a non-UTF8 database, character set translation was not

src/pgtclId.h
/*-------------------------------------------------------------------------
 *
 * pgtclId.h
 *
 *	Contains Tcl "channel" interface routines, plus useful routines
 *	to convert between strings and pointers.  These are needed because
 *	everything in Tcl is a string, but in C, pointers to data structures
 *	are needed.
 *
 * Portions Copyright (c) 1996-2004, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * $Id: pgtclId.h,v 1.2 2007/07/06 02:17:41 lbayuk Exp $
 *
 *-------------------------------------------------------------------------
 */

extern void PgSetConnectionId(Tcl_Interp *interp, PGconn *conn);

#define DRIVER_OUTPUT_PROTO ClientData cData, CONST84 char *buf, int bufSize, \
	int *errorCodePtr
#define DRIVER_INPUT_PROTO ClientData cData, char *buf, int bufSize, \
	int *errorCodePtr
#define DRIVER_DEL_PROTO ClientData cData, Tcl_Interp *interp

extern PGconn *PgGetConnectionId(Tcl_Interp *interp, CONST84 char *id,
				  Pg_ConnectionId **);
extern int	PgDelConnectionId(DRIVER_DEL_PROTO);
extern int	PgOutputProc(DRIVER_OUTPUT_PROTO);
extern int	PgInputProc(DRIVER_INPUT_PROTO);
extern int	PgSetResultId(Tcl_Interp *interp, CONST84 char *connid,
			  PGresult *res);
extern PGresult *PgGetResultId(Tcl_Interp *interp, CONST84 char *id);
extern void PgDelResultId(Tcl_Interp *interp, CONST84 char *id);
extern int	PgGetConnByResultId(Tcl_Interp *interp, CONST84 char *resid);
extern void PgStartNotifyEventSource(Pg_ConnectionId * connid);
extern void PgStopNotifyEventSource(Pg_ConnectionId * connid, char allevents);
extern void PgNotifyTransferEvents(Pg_ConnectionId * connid);
extern void PgConnLossTransferEvents(Pg_ConnectionId * connid);
extern void PgNotifyInterpDelete(ClientData clientData, Tcl_Interp *interp);

extern void PgClearResultCallback(Pg_ConnectionId *conn);

extern Tcl_ChannelType Pg_ConnType;


src/pgtclCmds.h
/*-------------------------------------------------------------------------
 *
 * pgtclCmds.h
 *	  declarations for the C functions which implement pg_* tcl commands
 *
 * Portions Copyright (c) 1996-2004, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * $Id: pgtclCmds.h 303 2011-09-19 00:55:11Z lbayuk $
 *
 *-------------------------------------------------------------------------
 */

#ifndef PGTCLCMDS_H
#define PGTCLCMDS_H

#include <tcl.h>

#include "libpq-fe.h"

/* Hack to deal with Tcl 8.4 const-ification without losing compatibility */
#ifndef CONST84
#define CONST84
#endif

#ifndef FALSE
#define FALSE 0
#endif
#ifndef TRUE
#define TRUE 1
#endif

#define RES_HARD_MAX 128
#define RES_START 16

/* Runtime Tcl version, set in Pgtcl_Init(), and used in some commands */
extern double pgtcl_tcl_version;

/*
 * Each Pg_ConnectionId has a list of Pg_TclNotifies structs, one for each
 * Tcl interpreter that has executed any pg_listens on the connection.
 * We need this arrangement to be able to clean up if an interpreter is
 * deleted while the connection remains open.  A free side benefit is that
 * multiple interpreters can be registered to listen for the same notify
 * name.  (All their callbacks will be called, but in an unspecified order.)
 *
 * We use the same approach for pg_on_connection_loss callbacks, but they
 * are not kept in a hashtable since there's no name associated.
 */

typedef struct Pg_TclNotifies_s
{
	struct Pg_TclNotifies_s *next;		/* list link */
	Tcl_Interp *interp;			/* This Tcl interpreter */

	/*
	 * NB: if interp == NULL, the interpreter is gone but we haven't yet
	 * got round to deleting the Pg_TclNotifies structure.
	 */
	Tcl_HashTable notify_hash;	/* Active pg_listen requests */

	char	   *conn_loss_cmd;	/* pg_on_connection_loss cmd, or NULL */
}	Pg_TclNotifies;

/*
 * Hash table entries in notify_hash of Pg_TclNotifies have the following
 * structure as their value. (Through 1.9.0, before the -pid option was
 * added to pg_listen,  the hash table entry values were just the callback
 * command string.)
 */
typedef struct Pg_notify_command_s
{
	char	   *callback;		/* Callback command, dynamically allocated */
	char		use_pid;		/* If non-zero, pass PID argument */
} Pg_notify_command;

typedef struct Pg_ConnectionId_s
{
	char		id[32];
	PGconn	   *conn;
	int			res_max;		/* Max number of results allocated */
	int			res_hardmax;	/* Absolute max to allow */
	int			res_count;		/* Current count of active results */
	int			res_last;		/* Optimize where to start looking */
	int			res_copy;		/* Query result with active copy */
	int			res_copyStatus; /* Copying status */
	PGresult  **results;		/* The results */

	Pg_TclNotifies *notify_list;	/* head of list of notify info */
	int			notifier_running;		/* notify event source is live */
	Tcl_Channel notifier_channel;		/* Tcl_Channel on which notifier
										 * is listening */
	char	   *null_string;	/* String to return for NULL values */
	Tcl_Obj	   *notice_command;	/* Command to execute on getting a NOTICE */
	Tcl_Interp *interp;			/* Pointer to Tcl interpreter. This is
								   used by the notice handler callback. */
	char	   *copyBuf;		/* Buffer, if COPY TO record is too big */
	char	   *copyBufNext;	/* COPY TO overflow buffer remaining data */
	int			copyBufLeft;	/* COPY TO overflow buffer remaining bytes */
	Tcl_Obj    *callbackPtr;    /* Callback command prefix for async queries */
	Tcl_Interp *callbackInterp; /* Interp for the async query callback */
}	Pg_ConnectionId;

/* Values of res_copyStatus */
#define RES_COPY_NONE	0
#define RES_COPY_INPROGRESS 1
#define RES_COPY_FIN	2


/* Compatibility macro for Tcl-8.6 which deprecates use of fields inside
   a Tcl_Interp. This uses a function which is new at Tcl-8.6 to get the
   errorLine value from inside the interpreter.
*/
#if (TCL_MAJOR_VERSION > 8) || (TCL_MAJOR_VERSION == 8 && TCL_MINOR_VERSION >= 6)
#define Get_ErrorLine(interp) Tcl_GetErrorLine(interp)
#else
#define Get_ErrorLine(interp) interp->errorLine
#endif


/* **************************/
/* registered Tcl functions */
/* **************************/
extern int Pg_conndefaults(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_connect(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_disconnect(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_exec(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_execute(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_select(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_result(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_open(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_close(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_read(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_write(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_lseek(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_creat(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_tell(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_unlink(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_import(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_lo_export(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_listen(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_sendquery(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_sendquery_params(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_sendquery_prepared(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_getresult(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_isbusy(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_blocking(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_cancelrequest(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_on_connection_loss(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_escape_string(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_quote(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_escape_bytea(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_unescape_bytea(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_transaction_status(
  ClientData cdata, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_parameter_status(
  ClientData cdata, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_exec_prepared(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_exec_params(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_notice_handler(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_result_callback(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

#ifdef HAVE_PQENCRYPTPASSWORD /* PostgreSQL >= 8.2.0 */
extern int Pg_encrypt_password(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);
#endif

#ifdef HAVE_LO_TRUNCATE /* PostgreSQL >= 8.3.0 */
extern int Pg_lo_truncate(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);
#endif

#ifdef HAVE_PQDESCRIBEPREPARED /* PostgreSQL >= 8.2.0 */
extern int Pg_describe_cursor(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);
extern int Pg_describe_prepared(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);
#endif

extern int Pg_backend_pid(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

extern int Pg_server_version(
  ClientData cData, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]);

#endif   /* PGTCLCMDS_H */


src/extra/pgtkAppInit.c
/*
 * pgtkAppInit.c
 *    $Id: pgtkAppInit.c,v 1.1 2004/02/29 20:54:55 lbayuk Exp $
 *    A Tcl/Tk shell ("wish") with PostgreSQL interface commands
 *
 * Portions Copyright (c) 1996-2003, PostgreSQL Global Development Group
 * Portions Copyright (c) 1993 The Regents of the University of California.
 * Copyright (c) 1994 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 */

#include <tk.h>
#include "libpgtcl.h"

/*
 *----------------------------------------------------------------------
 *
 * main
 *
 *		This is the main program for the application.
 *
 * Results:
 *		None: Tk_Main never returns here, so this procedure never
 *		returns either.
 *
 * Side effects:
 *		Whatever the application does.
 *
 *----------------------------------------------------------------------
 */

int
main(int argc, char **argv)
{
	extern int Tcl_AppInit (Tcl_Interp *interp);

	Tk_Main(argc, argv, Tcl_AppInit);
	return 0;					/* Needed only to prevent compiler
								 * warning. */
}


/*
 *----------------------------------------------------------------------
 *
 * Tcl_AppInit
 *
 *		This procedure performs application-specific initialization.
 *		Most applications, especially those that incorporate additional
 *		packages, will have their own version of this procedure.
 *
 * Results:
 *		Returns a standard Tcl completion code, and leaves an error
 *		message in interp->result if an error occurs.
 *
 * Side effects:
 *		Depends on the startup script.
 *
 *----------------------------------------------------------------------
 */

int
Tcl_AppInit(Tcl_Interp *interp)
{
	if (Tcl_Init(interp) == TCL_ERROR)
		return TCL_ERROR;
	if (Tk_Init(interp) == TCL_ERROR)
		return TCL_ERROR;

	/*
	 * Call the init procedures for libpgtcl.
	 */
	if (Pgtcl_Init(interp) == TCL_ERROR)
		return TCL_ERROR;

	/*
	 * Specify a user-specific startup file to invoke if the application
	 * is run interactively.  Typically the startup file is "~/.apprc"
	 * where "app" is the name of the application.	If this line is
	 * deleted then no user-specific startup file will be run under any
	 * conditions.
	 */

	Tcl_SetVar(interp, "tcl_rcFileName", "~/.wishrc", TCL_GLOBAL_ONLY);

	return TCL_OK;
}


src/extra/pgtclAppInit.c
/*
 * pgtclAppInit.c
 *    $Id: pgtclAppInit.c,v 1.1 2004/02/29 20:54:55 lbayuk Exp $
 *    A Tcl shell with PostgreSQL interface commands
 *
 * Portions Copyright (c) 1996-2003, PostgreSQL Global Development Group
 * Portions Copyright (c) 1993 The Regents of the University of California.
 * Copyright (c) 1994 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 */

#include <tcl.h>
#include "libpgtcl.h"

/*
 *----------------------------------------------------------------------
 *
 * main
 *
 *		This is the main program for the application.
 *
 * Results:
 *		None: Tcl_Main never returns here, so this procedure never
 *		returns either.
 *
 * Side effects:
 *		Whatever the application does.
 *
 *----------------------------------------------------------------------
 */

int
main(int argc, char **argv)
{
	extern int Tcl_AppInit (Tcl_Interp *interp);

	Tcl_Main(argc, argv, Tcl_AppInit);
	return 0;					/* Needed only to prevent compiler
								 * warning. */
}


/*
 *----------------------------------------------------------------------
 *
 * Tcl_AppInit
 *
 *		This procedure performs application-specific initialization.
 *		It calls libpgtcl to initialize its commands.
 *
 * Results:
 *		Returns a standard Tcl completion code, and leaves an error
 *		message in interp->result if an error occurs.
 *
 * Side effects:
 *		Depends on the startup script.
 *
 *----------------------------------------------------------------------
 */

int
Tcl_AppInit(Tcl_Interp *interp)
{
	if (Tcl_Init(interp) == TCL_ERROR)
		return TCL_ERROR;

	/*
	 * Call the init procedure for libpgtcl.
	 */
	if (Pgtcl_Init(interp) == TCL_ERROR)
		return TCL_ERROR;

	/*
	 * Specify a user-specific startup file to invoke if the application
	 * is run interactively.  Typically the startup file is "~/.apprc"
	 * where "app" is the name of the application.	If this line is
	 * deleted then no user-specific startup file will be run under any
	 * conditions.
	 */
	Tcl_SetVar(interp, "tcl_rcFileName", "~/.tclshrc", TCL_GLOBAL_ONLY);

	return TCL_OK;
}


src/pgtcl.c
/*-------------------------------------------------------------------------
 *
 * pgtcl.c
 *
 *	libpgtcl is a tcl package for front-ends to interface with PostgreSQL.
 *	It's a Tcl wrapper for libpq.
 *
 * Portions Copyright (c) 1996-2004, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  $Id: pgtcl.c 301 2011-09-19 00:55:01Z lbayuk $
 *
 *-------------------------------------------------------------------------
 */

#include "libpgtcl.h"
#include "pgtclCmds.h"
#include "pgtclId.h"

/* Runtime Tcl version, set below */
double pgtcl_tcl_version;

/*
 * Pgtcl_Init
 *	  initialization package for the PGTCL Tcl package
 *
 */

DLLEXPORT int
Pgtcl_Init(Tcl_Interp *interp)
{

	/*
	 * The version required really should be TCL_VERSION, but this
	 * can be compiled under 8.5 with stubs and still mostly works
	 * with a Tcl 8.4 interpreter, so let 8.4 be the minimum version.
	 */
#ifdef USE_TCL_STUBS
	if (Tcl_InitStubs(interp, "8.4", 0) == NULL)
		return TCL_ERROR;
#endif

	/*
	 * Get the Tcl version at runtime, which may differ from the compile-time
	 * version due to use of Tcl stubs. This is used in some commands to
	 * prevent crashes due to missing stubs functions.
	 */
	if (Tcl_GetDouble(interp,
				Tcl_GetVar(interp, "tcl_version", TCL_GLOBAL_ONLY),
				&pgtcl_tcl_version) != TCL_OK)
		pgtcl_tcl_version = 8.4;  /* A reasonable fallback */

	/*
	 * Note: Removed code to set PGCLIENTENCODING=UNICODE if tcl_version >= 8.1.
	 * That did not work for Windows because the libpq DLL didn't see the
	 * environment change. So now this is done when connecting to the database.
	 */

	/* register all pgtcl commands */
	Tcl_CreateObjCommand(interp,
						 "pg_conndefaults",
						 Pg_conndefaults,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_connect",
						 Pg_connect,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_disconnect",
						 Pg_disconnect,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_exec",
						 Pg_exec,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_select",
						 Pg_select,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_result",
						 Pg_result,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_execute",
						 Pg_execute,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_open",
						 Pg_lo_open,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_close",
						 Pg_lo_close,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_read",
						 Pg_lo_read,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_write",
						 Pg_lo_write,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_lseek",
						 Pg_lo_lseek,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_creat",
						 Pg_lo_creat,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_tell",
						 Pg_lo_tell,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_unlink",
						 Pg_lo_unlink,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_import",
						 Pg_lo_import,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_lo_export",
						 Pg_lo_export,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_listen",
						 Pg_listen,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_sendquery",
						 Pg_sendquery,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_sendquery_prepared",
						 Pg_sendquery_prepared,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_sendquery_params",
						 Pg_sendquery_params,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_getresult",
						 Pg_getresult,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_isbusy",
						 Pg_isbusy,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_blocking",
						 Pg_blocking,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						 "pg_cancelrequest",
						 Pg_cancelrequest,
						 (ClientData)NULL,
						 (Tcl_CmdDeleteProc *)NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_on_connection_loss",
						  Pg_on_connection_loss,
						  (ClientData) NULL, 
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_escape_string",
						  Pg_escape_string,
						  (ClientData) NULL, 
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_quote",
						  Pg_quote,
						  (ClientData) NULL, 
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_escape_bytea",
						  Pg_escape_bytea,
						  (ClientData) NULL, 
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_unescape_bytea",
						  Pg_unescape_bytea,
						  (ClientData) NULL, 
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_transaction_status",
						  Pg_transaction_status,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_parameter_status",
						  Pg_parameter_status,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_exec_prepared",
						  Pg_exec_prepared,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_exec_params",
						  Pg_exec_params,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_notice_handler",
						  Pg_notice_handler,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_result_callback",
						  Pg_result_callback,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

#ifdef HAVE_PQENCRYPTPASSWORD /* PostgreSQL >= 8.2.0 */
	Tcl_CreateObjCommand(interp,
						  "pg_encrypt_password",
						  Pg_encrypt_password,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);
#endif

#ifdef HAVE_LO_TRUNCATE /* PostgreSQL >= 8.3.0 */
	Tcl_CreateObjCommand(interp,
						  "pg_lo_truncate",
						  Pg_lo_truncate,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);
#endif

#ifdef HAVE_PQDESCRIBEPREPARED /* PostgreSQL >= 8.2.0 */
	Tcl_CreateObjCommand(interp,
						  "pg_describe_cursor",
						  Pg_describe_cursor,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_describe_prepared",
						  Pg_describe_prepared,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);
#endif

	Tcl_CreateObjCommand(interp,
						  "pg_backend_pid",
						  Pg_backend_pid,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	Tcl_CreateObjCommand(interp,
						  "pg_server_version",
						  Pg_server_version,
						  (ClientData) NULL,
						  (Tcl_CmdDeleteProc *) NULL);

	/* Note PACKAGE_VERSION (or VERSION) is provided by the TEA Makefile */
#ifndef PACKAGE_VERSION
#ifdef VERSION
#define PACKAGE_VERSION VERSION
#else
#define PACKAGE_VERSION "0.0"
#endif
#endif
	Tcl_PkgProvide(interp, "Pgtcl", PACKAGE_VERSION);

	return TCL_OK;
}

DLLEXPORT int
Pgtcl_SafeInit(Tcl_Interp *interp)
{
	return Pgtcl_Init(interp);
}
